src/sfizz/FilePool.cpp

```diff
@@ -617,57 +795,103 @@ void sfz::FilePool::setRamLoading(bool loadInRam) noexcept
+            fileData->status = FileData::Status::FullLoaded;
         }
     } else {
         setPreloadSize(preloadSize);
     }
 }
 
-void sfz::FilePool::triggerGarbageCollection() noexcept
+void sfz::FilePool::startRender()
 {
-    const std::unique_lock<SpinMutex> guard { garbageAndLastUsedMutex, std::try_to_lock };
-    if (!guard.owns_lock())
-        return;
+    globalObject->runningRender++;
+}
 
-    const auto now = std::chrono::high_resolution_clock::now();
-    swapAndPopAll(lastUsedFiles, [&](const FileId& id) {
-        if (garbageToCollect.size() == garbageToCollect.capacity())
-           return false;
+void sfz::FilePool::stopRender()
+{
+    globalObject->runningRender--;
+    std::error_code ec;
+    globalObject->semGarbageBarrier.post(ec);
+    ASSERT(!ec);
+}
 
-        auto it = preloadedFiles.find(id);
-        if (it == preloadedFiles.end()) {
-            // Getting here means that the preloadedFiles got changed (probably cleared)
-            // while the lastUsedFiles were untouched.
-            return true;
+void sfz::FilePool::GlobalObject::garbageJob()
+{
+    while (semGarbageBarrier.timed_wait(sfz::config::fileClearingPeriod * 1000), garbageFlag) {
+        if (runningRender != 0) {
+            continue;
         }
+        const auto now = highResNow();
+        const auto timeSinceLastCollection =
+        std::chrono::duration_cast<std::chrono::seconds>(now - lastGarbageCollection_);
 
-        sfz::FileData& data = it->second;
-        if (data.status == FileData::Status::Preloaded)
-            return true;
-
-        if (data.status != FileData::Status::Done)
-            return false;
-
-        if (data.readerCount != 0)
-            return false;
-
-        const auto secondsIdle = std::chrono::duration_cast<std::chrono::seconds>(now - data.lastViewerLeftAt).count();
-        if (secondsIdle < config::fileClearingPeriod)
-            return false;
-
-        data.availableFrames = 0;
-        data.status = FileData::Status::Preloaded;
-        garbageToCollect.push_back(std::move(data.fileData));
-        return true;
-    });
-
-    std::error_code ec;
-    semGarbageBarrier.post(ec);
-    ASSERT(!ec);
+        if (timeSinceLastCollection.count() < config::fileClearingPeriod) {
+            continue;
+        }
+        lastGarbageCollection_ = now;
+
+        {
+            std::unique_lock<std::mutex> guard { preloadedFilesMutex, std::try_to_lock };
+            if (guard.owns_lock()) {
+                for (auto it = preloadedFiles.begin(); it != preloadedFiles.end();) {
+                    auto copyIt = it++;
+                    auto& data = copyIt->second;
+
+                    if (data->canRemove()) {
+                        preloadedFiles.erase(copyIt);
+                        continue;
+                    }
+
+                    if (data->availableFrames == 0 || data->readerCount != 0)
+                        continue;
+
+                    switch (FileData::Status(data->status)) {
```

Comment:
Maybe this could be an `if` statement for now.

Reply:
data->status is std::atomic value, so you should define a variable when we use an `if` statement.

Reply:
I see I missed this, thanks.